internal/util/sizeof_test.go

```diff
@@ -0,0 +1,376 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"crypto/md5"
+	"net/http"
+	"testing"
+	"time"
+	"unsafe"
+
+	"github.com/googlecloudplatform/gcsfuse/internal/storage/gcs"
+	. "github.com/jacobsa/ogletest"
+	"google.golang.org/api/googleapi"
+	storagev1 "google.golang.org/api/storage/v1"
+)
+
+func TestSizeof(t *testing.T) { RunTests(t) }
+
+////////////////////////////////////////////////////////////////////////
+// Boilerplate
+////////////////////////////////////////////////////////////////////////
+
+type SizeofTest struct {
+}
+
+func init() { RegisterTestSuite(&SizeofTest{}) }
+
+var (
+	i            int
+	intArray     []int
+	b            byte
+	emptyString  string = ""
+	stringArray  []string
+	stringIntMap map[string]int
+
+	sizeOfInt               int
+	sizeOfIntPtr            int
+	sizeOfByte              int
+	sizeOfEmptyString       int
+	sizeOfEmptyIntArray     int
+	sizeOfEmptyStringArray  int
+	sizeOfEmptyStringIntMap int
+	sizeOfEmptyStruct       int
+	sizeOfEmptyGcsObject    int
+	sizeOfEmptyMinObject    int
+)
+
+func init() {
+	type emptyStruct struct{}
+
+	sizeOfInt = int(unsafe.Sizeof(i))
+	sizeOfIntPtr = int(unsafe.Sizeof(&i))
+	sizeOfByte = int(unsafe.Sizeof(b))
+	sizeOfEmptyString = int(unsafe.Sizeof(emptyString))
+	sizeOfEmptyIntArray = int(unsafe.Sizeof(intArray))
+	sizeOfEmptyStringArray = int(unsafe.Sizeof(stringArray))
+	sizeOfEmptyStringIntMap = int(unsafe.Sizeof(stringIntMap))
+
+	sizeOfEmptyStruct = int(unsafe.Sizeof(emptyStruct{}))
+	AssertEq(0, sizeOfEmptyStruct)
+
+	sizeOfEmptyGcsObject = int(unsafe.Sizeof(gcs.Object{}))
+	sizeOfEmptyMinObject = int(unsafe.Sizeof(gcs.MinObject{}))
+}
+
+////////////////////////////////////////////////////////////////////////
+// Tests
+////////////////////////////////////////////////////////////////////////
+
+func (t *SizeofTest) TestUnsafeSizeOf() {
+
+	for _, tc := range []struct {
+		t             any
+		expected_size int
+	}{
+		{
+			t:             i,
+			expected_size: sizeOfInt,
+		},
+		{
+			t:             &i,
+			expected_size: sizeOfIntPtr,
+		},
+		{
+			t: struct {
+			}{},
+			expected_size: sizeOfEmptyStruct,
+		},
+		{
+			t: struct {
+				x int
+			}{},
+			expected_size: sizeOfEmptyStruct + sizeOfInt,
+		},
+		{
+			t: struct {
+				a          int
+				b1, b2, b3 byte
+				c          string
+			}{},
+			expected_size: sizeOfEmptyStruct + sizeOfInt + 3*sizeOfByte + 5 /*for-padding-for-alignment*/ + sizeOfEmptyString,
```

Comment:
I didn't get why 5?

Reply:
5 is what compiler pads between b3 and c to align c to a multiple of 8 (a is at offset 0, b1 at 8, b2 at 9, b3 at 10, c at 16). This is a compiler feature. Compilers always need to align structures within structures at offsets which are multiples of the platform-int-size (8 bytes in this case).